Sources/XcbeautifyLib/JunitReporter.swift

```diff
@@ -0,0 +1,184 @@
+// Information about the JUNIT Schema/specification used in this file can be found here:
+// * https://stackoverflow.com/a/9410271
+// * https://github.com/bazelbuild/bazel/blob/45092bb122b840e3410845522df9fe89c59db465/src/java_tools/junitrunner/java/com/google/testing/junit/runner/model/AntXmlResultWriter.java#L29
+// * http://windyroad.com.au/dl/Open%20Source/JUnit.xsd
+
+import Foundation
+import XMLCoder
+
+public final class JunitReporter {
+    private var components: [JunitComponent] = []
+    
+    public init() { }
+
+    public func add(line: String) {
+        switch line {
+            case Matcher.failingTestMatcher:
+                components.append(.failingTest(line))
+            case Matcher.testCasePassedMatcher:
+                components.append(.testCasePassed(line))
+        case Matcher.testSuiteStartMatcher:
+                components.append(.testSuiteStart(line))
+            default:
+                // Not needed for generating a junit report
+                break
+        }
+    }
+    
+    public func generateReport() throws -> Data {
+        let parser = JunitComponentParser()
+        components.forEach { parser.parse(component: $0) }
+        let encoder = XMLEncoder()
+        encoder.keyEncodingStrategy = .lowercased
+        encoder.outputFormatting = [.prettyPrinted]
+        let result = parser.result()
+        return try encoder.encode(result)
+    }
+}
+
+private final class JunitComponentParser {
+    private var mainTestSuiteName: String?
+    private var testCases: [Testcase] = []
+
+    func parse(component: JunitComponent) {
+        switch component {
+        case let .testSuiteStart(line):
+            guard mainTestSuiteName == nil else {
+                break
+            }
+            let groups = line.capturedGroups(with: .testSuiteStart)
+            mainTestSuiteName = groups[0]
+
+        case let .failingTest(line):
+            let groups = line.capturedGroups(with: .failingTest)
+            let testCase = Testcase(
+                classname: groups[1],
+                name: groups[2],
+                time: nil,
+                failure: .init(message: "\(groups[0]) - \(groups[3])")
+            )
+            testCases.append(testCase)
+
+        case let .testCasePassed(line):
+            let groups = line.capturedGroups(with: .testCasePassed)
+            let testCase = Testcase(classname: groups[0], name: groups[1], time: groups[2], failure: nil)
+            testCases.append(testCase)
+        }
+    }
+
+    func result() -> Testsuites {
+        var groupedTestCases: [String: [Testcase]] = [:]
+        for testCase in testCases {
+            groupedTestCases[testCase.classname] = (groupedTestCases[testCase.classname] ?? []) + [testCase]
+        }
+        let testSuites = groupedTestCases.map { Testsuite(name: $0.key, testcases: $0.value) }
+        let container = Testsuites(name: mainTestSuiteName, testsuites: testSuites)
+        return container
```

Comment:
It is not fully clear to me the specification for junit reports, but based on https://stackoverflow.com/a/9410271 seems like junit reports have 2 nested levels, and testsuites can only have testcases inside (not other testsuites):
```
<testsuites ...>
    <testsuite ...>
        <testcase ... > </testcase>
        <testcase ... > </testcase>
    </testsuite>
        <testcase ... > </testcase>
        <testcase ... > </testcase>
    </testsuite>
</testsuites>
```

Thus, the idea is to collect all testCases, and group them together in suites based on their `classname`, so there is no need to rely on the order of the lines, making the execution more robust